### PR TITLE
Compact header

### PR DESCRIPTION
**Pull request by maykar**
## Proposed change

This compacts the header to combine both buttons and tabs. Additionally, this modifies views so their backgrounds extend behind the header in Lovelace. When using a transparent color or image for the header background the view's background will now show underneath.

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information

compactheader

<hr>

#### @ 320px wide

image

## Checklist

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

**Comment by zsarnett** [comment]
ITS FINALLY HAPPENING 🎉 🎉 🎉 🎉

File: src/panels/media-browser/ha-panel-media-browser.ts [hunk 1]
```diff
@@ -137,14 +137,15 @@ class PanelMediaBrowser extends LitElement {
           --mdc-theme-primary: var(--app-header-text-color);
         }
         ha-media-player-browse {
-          height: calc(100vh - 64px);
+          height: calc(100vh - 48px);
```
Comment (balloob) [inline src/panels/media-browser/ha-panel-media-browser.ts:140]: All these calculation changes seem weird, since the compact header is configurable. How will this work?

File: src/state/compact-header.mixin.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  interface HASSDomEvents {
+    "hass-compact-header": CompactHeaderParams;
+  }
+  // for add event listener
+  interface HTMLElementEventMap {
+    "hass-compact-header": HASSDomEvent<CompactHeaderParams>;
+  }
+}
+
+export default <T extends Constructor<HassBaseEl>>(superClass: T) =>
```
Comment (balloob) [inline src/state/compact-header.mixin.ts:22]: This does not need to be it's own mixin

File: src/state/hass-element.ts [hunk 3]
```diff
@@ -31,4 +32,5 @@ export class HassElement extends ext(HassBaseEl, [
   hapticMixin,
   panelTitleMixin,
   loggingMixin,
+  compactHeaderMixin,
```
Comment (balloob) [inline src/state/hass-element.ts:35]: This is now also added to Home Assistant Cast. We should not add it here.

**Comment by balloob** [comment]
How will this work when the user has more tabs that fit the view? If you have found a solution for that that works, there would no need to be an option and we can always do it.

File: src/panels/media-browser/ha-panel-media-browser.ts (hunk 1 above)
Comment (maykar) [reply src/panels/media-browser/ha-panel-media-browser.ts:140]: For anything that isn't in Lovelace, the calculations needed changed to fit the new size of app-header. For anything in Lovelace I still need to do an adjustment.  Both `hui-masonry-view` and `hui-panel-view` need to change the calculation depending on compact mode.

**Comment by maykar** [comment]
[My previous PR](https://github.com/home-assistant/frontend/pull/7238) added adjustments to paper-tabs in order to make scrolling work in such a way that the scroll buttons no longer take up extra space.

File: src/state/compact-header.mixin.ts (hunk 2 above)
Comment (maykar) [reply src/state/compact-header.mixin.ts:22]: I'm still uncertain on mixins and the best way to add this, will need some guidance/feedback.

**Comment by maykar** [comment]
Converting to draft as there are issues/questions with adding this as a profile preference and need to find the best way to adjust masonry and panel view dependent on compact mode.

File: src/panels/media-browser/ha-panel-media-browser.ts (hunk 1 above)
Comment (maykar) [reply src/panels/media-browser/ha-panel-media-browser.ts:140]: Since this is now a UI overhaul, there is no need to adjust masonry view or panel view. I will look at other areas that are using hui-root/app-header to see where adjustments might need made.

Comment (maykar) [reply src/panels/media-browser/ha-panel-media-browser.ts:140]: So it looks like the only UIs that would be affected would be the ones that actually use tabs: dev-tools panel, mailbox panel, and Lovelace. They all seem to already handle the change.

Comment (balloob) [reply src/panels/media-browser/ha-panel-media-browser.ts:140]: I wonder if we can make it all easier and just define the height of the toolbar in a var?

File: src/components/ha-tabs.ts [hunk 4]
```diff
@@ -34,6 +34,9 @@ export class HaTabs extends PaperTabs {
 
       superStyle!.appendChild(
         document.createTextNode(`
+          :host {
+            padding-top: .5px;
```
Comment (balloob) [inline src/components/ha-tabs.ts:38]: Spot the difference 🔎 😄

File: src/resources/styles.ts [hunk 5]
```diff
@@ -110,6 +110,11 @@ export const haStyle = css`
     color: var(--app-header-text-color, white);
   }
 
+  app-toolbar,
+  div[sticky] {
```
Comment (balloob) [inline src/resources/styles.ts:114]: This style is too generic and should be more specific.

File: src/components/ha-tabs.ts (hunk 4 above)
Comment (maykar) [reply src/components/ha-tabs.ts:38]: Oh trust me, it's visible. 😄

Without the .5px there is a slight line below the tab indicator, with a full 1px it overlaps the header bottom.

File: src/resources/styles.ts (hunk 5 above)
Comment (maykar) [reply src/resources/styles.ts:114]: How so? Both should always have that height or is it just the `div[sticky]` that is the issue? If that's the case, could do something like:

```
  app-toolbar,
  app-header div[sticky] {
    height: 48px;
  }
```

File: src/panels/media-browser/ha-panel-media-browser.ts (hunk 1 above)
Comment (maykar) [reply src/panels/media-browser/ha-panel-media-browser.ts:140]: Yeah, maybe like:```
--view-height-no-tabs: calc(100vh - 48px);
--view-height-tabs: calc(100vh - 96px);
```

File: src/resources/styles.ts (hunk 5 above)
Comment (maykar) [reply src/resources/styles.ts]: Added the above

**Comment by maykar** [comment]
Not quite done styling yet, but it's getting there.

**Comment by maykar** [comment]
Okay, pretty happy where it is ATM. Let me know if you'd like any styling changes or if anything seems off.

File: src/panels/media-browser/ha-panel-media-browser.ts (hunk 1 above)
Comment (maykar) [reply src/panels/media-browser/ha-panel-media-browser.ts:140]: Thinking about this more, the header itself can only really be 2 heights 48px or 96px depending if tabs are used or not, but the calculation that views below the header would use could be completely different then what I have above.

While the current files changed in this PR just use the single header value of 48px, that could change in the future. So rather than storing a calculation as the vars, it seems to make more sense to store just the single pixel values (48px, 96px) and at that point  `calc(100vh - 48px)` seems cleaner than something like `calc(100vh - var(--header-no-tabs))` though slightly less "readable".

Should we go ahead and add something like below or is that overkill? ```
--header-no-tabs: 48px
--header-with-tabs: 96px
```

Comment (maykar) [reply src/panels/media-browser/ha-panel-media-browser.ts:140]: Going to think on this and probably work it into a different PR as this one is just modifying already existing calculations. I think adding more vars than just ones that use `app-header`'s height would be useful.

File: src/panels/lovelace/views/hui-masonry-view.ts [hunk 6]
```diff
@@ -261,9 +261,10 @@ export class MasonryView extends LitElement implements LovelaceViewElement {
       :host {
         display: block;
         background: var(--lovelace-background);
-        padding-top: 4px;
-        height: 100%;
+        padding-top: 100px;
+        height: calc(100% + 160px);
         box-sizing: border-box;
+        margin-top: -96px;
```
Comment (bramkragten) [inline src/panels/lovelace/views/hui-masonry-view.ts:267]: Why is this needed?

Comment (maykar) [reply src/panels/lovelace/views/hui-masonry-view.ts:267]: This styling is what makes the view's background extend behind the header. Without this if you were to use a transparent background on app-header it would show empty space behind it rather than the view's background.

Comment (balloob) [reply src/panels/lovelace/views/hui-masonry-view.ts:267]: This feels very weird. We should just apply the background on the element that contains the tabs + view instead of mushing the view behind the tab bar. And instead of `margin-top` one should just go then for position absolute and just tag it to top, left, right, bottom all 0. But really, just setting the background on the parent is the best approach.

Comment (maykar) [reply src/panels/lovelace/views/hui-masonry-view.ts:267]: So, we should just be setting the background on `ha-app-layout` and dropping the styling changes I made on masonry-view and panel-view.

Comment (maykar) [reply src/panels/lovelace/views/hui-masonry-view.ts:267]: Sorry, guess I'm still thinking of things a bit in terms of a custom addon as this is exactly something I'd do for CH.

Comment (maykar) [reply src/panels/lovelace/views/hui-masonry-view.ts]: Removed background and styling changes from `hui-masonry-view` and `hui-panel-view` and added the background to `ha-app-layout`.

File: src/resources/styles.ts [hunk 7]
```diff
@@ -110,6 +110,11 @@ export const haStyle = css`
     color: var(--app-header-text-color, white);
   }
 
+  app-toolbar,
```
Comment (balloob) [inline src/resources/styles.ts:113]: This is ok for now. One day we should rip all these styles out of the "generic styles" page.

**Comment by balloob** [comment]
I miss updates to `hass-subpage.ts`, `hass-tabs-subpage.ts`, `hass-tabs-subpage-data-table.ts`, `hass-loading-screen.ts`, `hass-error-screen.ts`

Also let's make the height a variable so we don't have to hardcode 48 everywhere.

**Comment by bramkragten** [comment]
> Also let's make the height a variable so we don't have to hardcode 48 everywhere.

Then we can also make the header 48px on mobile but 65px on desktop

File: src/panels/lovelace/views/hui-masonry-view.ts [hunk 8]
```diff
@@ -260,7 +260,6 @@ export class MasonryView extends LitElement implements LovelaceViewElement {
     return css`
       :host {
         display: block;
-        background: var(--lovelace-background);
```
Comment (bramkragten) [inline src/panels/lovelace/views/hui-masonry-view.ts:263]: We allow a background per view, so we set the css variable of the right element now?

File: src/panels/lovelace/hui-root.ts [hunk 9]
```diff
@@ -675,7 +706,10 @@ class HUIRoot extends LitElement {
     const configBackground = viewConfig.background || this.config.background;
 
     if (configBackground) {
-      view.style.setProperty("--lovelace-background", configBackground);
+      root.parentElement!.style.setProperty(
```
Comment (bramkragten) [inline src/panels/lovelace/hui-root.ts:709]: This is dangerous, if the dom structure changes this will break. Change the query selectors to the decorator style and cache them. You can then just query the `#layout` element.

Als is setting the background for unused entities line 677-681 still needed?

**Comment by balloob** [comment]
That's a good point Bram. Let's not do it all in 1 PR and just focus on collapsing the tab bar into the header.

File: src/panels/lovelace/hui-root.ts (hunk 9 above)
Comment (maykar) [reply src/panels/lovelace/hui-root.ts:709]: I'll drop all extra styling for now like balloob suggests below, but I was thinking this too, since unused entities changed and is now in its own list box there's really no need for it either.

File: src/resources/ha-style.ts [hunk 10]
```diff
@@ -48,6 +48,9 @@ documentContainer.innerHTML = `<custom-style>
       --primary-background-color: #fafafa;
       --secondary-background-color: #e5e5e5; /* behind the cards on state */
 
+      /* for header */
+      --header-height: 48px;
```
Comment (bramkragten) [inline src/resources/ha-style.ts:52]: 65px

**Comment by MrJelly17** [comment]
So great to see someone working on this project, great job!

File: src/components/ha-tabs.ts [hunk 11]
```diff
@@ -34,14 +34,17 @@ export class HaTabs extends PaperTabs {
...
           .not-visible {
             display: none;
           }
-          :host > paper-icon-button:first-of-type {
-            padding-left: 0;
-          }
           paper-icon-button {
-            margin: 0 -8px 0 0;
+            width: 24px;
+            height: 48px;
```
Comment (balloob) [inline src/components/ha-tabs.ts:45]: Why are we introducing a `width` and `height` for paper-icon-button?

File: src/panels/lovelace/hui-root.ts [hunk 12]
```diff
@@ -554,10 +587,12 @@ class HUIRoot extends LitElement {
 
   private _enableEditMode(): void {
     this.lovelace!.setEditMode(true);
+    fireEvent(this, "iron-resize");
```
Comment (balloob) [inline src/panels/lovelace/hui-root.ts:590]: Does this work? Once we set editMode it will take a few event loop cycles before this element is updated. 

I think that we should put firing this event in `updated` and fire it whenever edit mode changes.

File: src/panels/lovelace/hui-root.ts [hunk 13]
```diff
@@ -697,13 +731,19 @@ class HUIRoot extends LitElement {
 
         ha-app-layout {
           min-height: 100%;
+          background: var(--lovelace-background);
         }
         ha-tabs {
-          margin-left: max(env(safe-area-inset-left), 24px);
-          margin-right: max(env(safe-area-inset-right), 24px);
+          width: 100%;
+          height: 100%;
```
Comment (balloob) [inline src/panels/lovelace/hui-root.ts:738]: Should this be
```suggestion
          height: var(--header-height);
```

File: src/panels/lovelace/hui-root.ts [hunk 14]
```diff
@@ -738,7 +778,7 @@ class HUIRoot extends LitElement {
           color: var(--error-color);
         }
         #view {
-          min-height: calc(100vh - 112px);
+          min-height: calc(100vh - var(--header-height));
```
Comment (balloob) [inline src/panels/lovelace/hui-root.ts:781]: This is weird? 112px is not the same as header height ?

File: src/panels/media-browser/ha-panel-media-browser.ts [hunk 15]
```diff
@@ -137,14 +137,15 @@ class PanelMediaBrowser extends LitElement {
           --mdc-theme-primary: var(--app-header-text-color);
         }
         ha-media-player-browse {
-          height: calc(100vh - 64px);
+          height: calc(100vh - var(--header-height));
```
Comment (balloob) [inline src/panels/media-browser/ha-panel-media-browser.ts:140]: If we're making these changes here, we should also make sure that we do this for all the layouts in `src/layouts`

File: src/components/ha-tabs.ts (hunk 11 above)
Comment (maykar) [reply src/components/ha-tabs.ts:45]: This is strictly for the scroll arrows in paper-tabs to allow for more space for tabs.

File: src/resources/styles.ts [hunk 16]
```diff
@@ -110,12 +110,16 @@ export const haStyle = css`
...
-  app-toolbar [main-title] {
-    margin-left: 20px;
+  app-toolbar {
+    height: var(--header-height);
   }
 
-  ha-menu-button {
-    margin-left: 4px;
+  app-header div[sticky] {
+    height: 48px;
```
Comment (balloob) [inline src/resources/styles.ts:118]: 48 or header-height ?

File: src/panels/lovelace/hui-root.ts (hunk 12 above)
Comment (maykar) [reply src/panels/lovelace/hui-root.ts:590]: This works well as is, without it going into edit mode causes the header to overlap and coming out causes the space taken up by the 2 rows to remain. Will make the change though.

File: src/resources/styles.ts (hunk 16 above)
Comment (maykar) [reply src/resources/styles.ts:118]: Sticky has always been 48, but when we changed to the variable and made app header 64px again it didn't get changed. It is correct at 48 now.

File: src/panels/lovelace/hui-root.ts (hunk 14 above)
Comment (maykar) [reply src/panels/lovelace/hui-root.ts:781]: 112px was incorrect and header height is the correction.

**Comment by TheDK** [comment]
Great to see this is becoming standard after Custom Header going away!

File: src/panels/media-browser/ha-panel-media-browser.ts (hunk 15 above)
Comment (maykar) [reply src/panels/media-browser/ha-panel-media-browser.ts:140]: Need to either adjust the header's, make another var, etc. for other layouts as they require the header's height + 1.

65px rather than 64px. Will check it out more later tonight or tomorrow.

**Comment by poldim** [comment]
@maykar will this make the header by default compact on all views from all devices? Or will it just now allow it to be adjusted via themes but default to non-compact style?

**Comment by bramkragten** [comment]
> @maykar will this make the header by default compact on all views from all devices? Or will it just now allow it to be adjusted via themes but default to non-compact style?

It will compact all Lovelace headers

File: src/panels/lovelace/hui-root.ts (hunk 13 above)
Comment (maykar) [reply src/panels/lovelace/hui-root.ts:733]: Ah, right. This should have stayed 100% and bram just fixed it. When in edit mode, ha-tabs ends up in div sticky. So it needs to be 100%

**Review by bramkragten — APPROVED** [review]

**Comment by RssFra97** [comment]
Can other changes be made? type transparent header with lovelace background? if yes, can it already be done in beta 1?

**Comment by bramkragten** [comment]
Please don't react to closed PR's
